### PR TITLE
Type aliases

### PR DESCRIPTION
**Pull request by flodiebold**
This implements type aliases (i.e. `type` definitions).

There's just one snag: handling recursion. E.g. `type Foo = Foo` makes type inference panic with a query cycle. I think the best way to handle this would be if Salsa provided the ability to catch cycle errors? It seems that there's some work underway to support this [here](https://github.com/salsa-rs/salsa/issues/6) and [here](https://github.com/salsa-rs/salsa/pull/147). Should we wait for this? I don't see a good way to handle this without help from Salsa.

### CODE REVIEW

File: crates/ra_hir/src/ty/lower.rs [hunk 1]
```diff
@@ -298,14 +302,24 @@ fn type_for_enum(db: &impl HirDatabase, s: Enum) -> Ty {
...
+    inner.subst(&substs)
+}
+
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
 pub enum TypableDef {
     Function(Function),
     Struct(Struct),
     Enum(Enum),
     EnumVariant(EnumVariant),
+    Type(Type),
```
Comment (matklad) [inline crates/ra_hir/src/ty/lower.rs]: Maybe let's call this `TypeAlias`? `Type` seems ambiguous.

**Comment by matklad** [comment]
FWIW, Kotlin seems to produce pretty specific errors in case of type alias recusion: https://github.com/JetBrains/kotlin/blob/ba6da7c40a6cc502508faf6e04fa105b96bc7777/compiler/frontend/src/org/jetbrains/kotlin/resolve/TypeAliasExpander.kt#L197

File: crates/ra_hir/src/ty/lower.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_hir/src/ty/lower.rs]: Ah, I guess the problem here is me: https://github.com/rust-analyzer/rust-analyzer/commit/0f9c350812e37b3cf2a1fcfb3c60a78122d60bf7#diff-8ce650d0905ec87e4d757e9b1aed9570R348

Should we rename `Type` to `TypeAlias` everywhere?

**Comment by flodiebold** [comment]
What I'm hoping for from Salsa is a way to say "if `type_for_def` is called recursively with the same type, just return `Ty::Unknown`". I think that would handle this in the most elegant way, and if it turns some bugs into just uninferred types, that's probably a plus ;) The other ways to handle this (like the two-phase expansion) add a lot of boilerplate for a corner case, I think.

File: crates/ra_hir/src/ty/lower.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir/src/ty/lower.rs]: Yeah... Though I'd rather do it as a separate PR :smile:

**Comment by flodiebold** [comment]
(About displaying a diagnostic and a quickfix etc., I imagine the `type_for_def` query would have to return something like `(Ty, Diagnostics)` for things like name resolution errors anyway, so then the cycle error would be returned in the `Diagnostics`...)

**Pull request merged**